features/268-next-cf/api.md

```diff
@@ -0,0 +1,309 @@
+- `config`: the function configuration
+- `status`: `function created`, `deploying`, `deploy failed`, `running`, `stopped`
+- `deployed_at`: time when the function is successfully deployed
+- `failed_message`: message or data of failed deployment from the k8s or faas platform
+- `plan`: pricing plan associated with the function, indicating how much resource would be allocated and the allowed idle time
+
+### Immutable deployment
+
+When function of the same name get deployed again, a new record of a version of the function would be created and the old function would not be deleted immediately. There would be a number of version of the same function deployed on OpenFaas.
+
+By default, the latest version of the function would be exposed to the endpoint as configured by the skygear user. When there are needs to access to older version, the request only need to includes the version hash to tell the CF Gateway to invoke the specific version. (Where and how to put the version hash can be further discussed, e.g. path or header)
+
+Depending on our pricing model and resource management on cloud, when new version of a function is deployed, the older version may be stopped to restrict concurrent functions. And older versions may be restricted to sleep in a shorter period of idle time.
+
+Skycli may have command to set the current version of a function to a specific version.
+
+#### App version
+
+For easier management, there may be a concept of app version, which represent a collection of function version within the same app.
+
+Each time developer deploy functions or remove functions, there will be a new app version created. CF Controller may have api for developer rollback all functions to a previous version by setting the app version to a previous one.
+
+*I think app version would be a simpler concept and a simpler thing to work with than function version for developers. Technically, we may support both at the same time or either one.*
+
+# Function configuration
+
+Function configuration may be placed in the file for other skycli configuration.
+
+Example:
+
+```yaml
+cf:
+  function1:
+    type: http-handler
+    path: /hello-world
+    env: node
+    src: js/hellow-world
+    secrets:
+      - DATABASE_URL
+    permission:
+      - name: key_required
+      - name: user_required
+```
+
+- `cf` indicates the dictionary is the cf configuration in the file
+- `function1` is the name of function, function deployed with the same name will considered as the same function with different version
+- `type` is the type of the function
+  - `http-handler`
+    - a single function that accepts http requests from an exact path
+    - anytime may be scaled down to zero
+  - `http-service`
+    - an http server that accepts http requests from the path
+    - may be long running
+  - `function`:
+    - a single function (in a program) that read input from stdin and write output to stdout
```

Comment:
May I know more about what is the stdin and stdout of a Cloud Function? When will I use it?

Reply:
`http-handler` and `http-service` are targeted to expose http api, however, we may need to support other types of trigger, e.g. cron job, event trigger.

stdin and stdout function is natively supported by Openfaas, so I think it's good to choose it as the base type for other trigger.

Otherwise if we want a long running daemon to serve these type of functions, we may consider using other ipc.